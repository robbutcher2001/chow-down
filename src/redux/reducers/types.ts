export interface Message {
    user: string
    message: string
}

export interface HelloState {
    hello_messages: Message[]
}

export const SAY_HELLO = 'SAY_HELLO'
export const DELETE_HELLO = 'DELETE_HELLO';

interface SayHelloAction {
    type: typeof SAY_HELLO
    rob_payload: Message
}

interface DeleteHelloAction {
    type: typeof DELETE_HELLO
}

export type SayHelloTypes = SayHelloAction | DeleteHelloAction