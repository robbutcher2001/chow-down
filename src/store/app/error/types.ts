import { Action } from 'redux';

export enum ErrorActionTypes {
  UNEXPECTED_SERVER_ERROR = '@@common/UNEXPECTED_SERVER_ERROR',
  UNEXPECTED_RESPONSE = '@@common/UNEXPECTED_RESPONSE',
  CLEAR_ERROR = '@@common/CLEAR_ERROR'
}

export interface ErrorState {
  readonly message: string
}

export interface ErrorMessageApiResponse {
  message: string
}

export interface ErrorApiResponse extends Action {
  type: ErrorActionTypes,
  json: ErrorMessageApiResponse
}