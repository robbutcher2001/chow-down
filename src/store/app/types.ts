import { Action } from 'redux';

export enum AppActionTypes {
    UNEXPECTED_SERVER_ERROR = '@@common/UNEXPECTED_SERVER_ERROR',
    UNEXPECTED_RESPONSE = '@@common/UNEXPECTED_RESPONSE',
    CLEAR_ERROR = '@@common/CLEAR_ERROR'
}

export interface AppState {
    readonly error: string
}

export interface ErrorMessageApiResponse {
    message: string
}

export interface ErrorApiResponse extends Action {
    type: AppActionTypes,
    json: ErrorMessageApiResponse
}