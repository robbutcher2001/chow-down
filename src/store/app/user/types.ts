import { Action } from 'redux';

export enum UserActionTypes {
  SET_USER_IS_SELECTING_DAY = '@@common/SET_USER_IS_SELECTING_DAY',
  CLEAR_USER_IS_SELECTING_DAY = '@@common/CLEAR_USER_IS_SELECTING_DAY'
}

export interface UserState {
  readonly selectedDay: string
}

export interface UserAction extends Action {
  type: UserActionTypes,
  day: string
}