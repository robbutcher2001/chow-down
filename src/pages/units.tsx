import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { GlobalState } from '../store';
import { UnitsState, GetUnitsApiRequest, PostUnitApiRequest } from '../store/domain/units/types';
import { getUnitsRequest, postUnitsRequest } from '../store/domain/units/actions';

import Form from '../components/form';
import InputBox from '../components/input-box';

interface DispatchProps {
    getUnits: () => GetUnitsApiRequest,
    postUnit: (form: object) => PostUnitApiRequest
};

interface OwnProps { };

interface OwnState { };

type CombinedProps = UnitsState & DispatchProps & OwnProps;

class UnitsPage extends Component<CombinedProps, OwnState> {
    constructor(props: CombinedProps) {
        super(props);
    }

    requestUnits = () => this.props.getUnits();
    addUnit = (form: object) => this.props.postUnit(form);

    componentDidMount = () => this.props.getUnits();

    render = () => {
        console.log(this.props.units);
        return (
            <div>
                <h4>List units</h4>
                <button onClick={this.requestUnits}>
                    Press me to get all units
                </button>
                {/* {this.props.pending &&
                    <div style={{ color: 'red' }}>Getting..</div>
                } */}
                {this.props.units.map((unit, index) =>
                    <li key={index}>{unit.singular}, {unit.plural}</li>
                )}
                <h4>Add unit</h4>
                <Form
                    dispatch={this.addUnit}
                    submitText='Add unit'>
                    <InputBox
                        name='singular'
                        placeholderText='Singular unit name'
                    />
                    <InputBox
                        name='plural'
                        placeholderText='Plural unit name'
                    />
                </Form>
            </div>
        );
    }
};

const mapStateToProps = ({ domain }: GlobalState, ownProps: OwnProps): UnitsState => ({
    error: domain.unit.error,
    units: domain.unit.units
});

const mapDispatchToProps = (dispatch: Dispatch, ownProps: OwnProps): DispatchProps => ({
    getUnits: () => dispatch(getUnitsRequest()),
    postUnit: (form: object) => dispatch(postUnitsRequest(form))
});

export default connect<UnitsState, DispatchProps, OwnProps, GlobalState>
    (mapStateToProps, mapDispatchToProps)(UnitsPage);