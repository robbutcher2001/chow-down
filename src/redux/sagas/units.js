import { call, put, takeLatest } from 'redux-saga/effects';

import { Method } from '../../globals/constants';
import api from '../../globals/api';

const URL = 'http://localhost:3000/api/units';

export default function* watcherSaga() {
    // yield takeLatest('GET_UNITS_REQUEST', test);
    yield takeLatest('POST_UNIT_REQUEST', workerSaga);
}

function* test() {
    try {
        yield put({ type: 'GET_UNITS_REQUEST_PENDING' });

        const payload = yield call(() => fetch(URL).then(data => data.json()));
        yield put({
            type: 'GET_UNITS_SUCCESS',
            payload
        });
    } catch (error) {
        console.error(error);
    }
}

function* workerSaga({ payload }) {
    try {
        yield put({ type: 'POST_UNIT_REQUEST_PENDING' });

        const response = yield call(() => api(
            Method.POST,
            URL,
            payload
        ));

        if (response.status === 200) {
            const json = yield response.json();
            yield put({
                type: 'POST_UNIT_SUCCESS',
                payload: json
            });
            yield put({ type: 'GET_UNITS_REQUEST' });
        }
        else {
            throw response;
        }
    } catch (err) {
        if (err.status >= 400 && err.status < 500) {
            const json = yield err.json();
            yield put({
                type: 'POST_UNIT_FAILED',
                payload: json
            });

            return;
        }

        //TODO: migrate this so what ever this returns, the reducer gives a useful message back to the components
        yield put({
            type: 'POST_UNIT_FAILED',
            payload: err.statusText
        });
    }
}