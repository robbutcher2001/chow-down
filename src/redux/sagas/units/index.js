'use strict';

import { all, takeLatest } from 'redux-saga/effects';

import { Actions } from '../../../globals/constants';

import getSaga from './get';
import postSaga from './post';

export default function* watcherSaga() {
    console.log('[saga] Started units saga');
    yield all([
        takeLatest(Actions.units.GET_UNITS_REQUEST, getSaga),
        takeLatest(Actions.units.POST_UNIT_REQUEST, postSaga)
    ]);
}