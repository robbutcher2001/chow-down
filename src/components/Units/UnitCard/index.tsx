import React from 'react';

import styled from 'styled-components';

import { Unit } from '../../../store/domain/units/types';

interface UnitCardProps {
  unit: Unit
};

const UnitCard = styled.li`
  background: rgba(171, 184, 195, 0.25);
  border-left-style: solid;
  border-left-width: 0.35rem;
  border-left-color: #0693E3;
  margin: 0.5rem;
  padding: 0.75rem;

  aside {
    font-size: 0.8rem;
    font-style: italic;
    margin-top: 0.5rem;

    &::before {
      content: "("
    }

    &::after {
      content: ")"
    }
  }
`

export default (props: UnitCardProps) => (
  <UnitCard>
    {props.unit.singular}
    <aside>{props.unit.plural}</aside>
  </UnitCard>
);