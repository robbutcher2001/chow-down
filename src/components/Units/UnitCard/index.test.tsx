import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import UnitCard from '.';
import { Unit } from '../../../store/domain/units/types';

const mockUnit: Unit = {
  id: '123',
  singular: 'fake_singular',
  plural: 'fake_plural'
};

test('UnitCard basic snapshot render', () => {
  const unitCard = renderer.create(
    <UnitCard unit={mockUnit} />
  );

  expect(unitCard.toJSON()).toMatchSnapshot();
});

test('UnitCard correct tag content assertion', () => {
  const { getByText } = render(
    <UnitCard unit={mockUnit} />
  );

  expect(getByText(/fake_singular/).textContent).toEqual('fake_singularfake_plural');
});