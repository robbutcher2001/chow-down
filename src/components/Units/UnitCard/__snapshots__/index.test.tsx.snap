// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitCard basic snapshot render 1`] = `
<li
  className="sc-AxjAm iNhScg"
>
  fake_singular
  <aside>
    fake_plural
  </aside>
</li>
`;
